### PR TITLE
[FEATURE] Add StreamingRandomPatchesClassifier ensemble method

### PR DESCRIPTION
**Pull request by jacobmontiel**
Changes proposed in this pull request:

Add the Streaming Random Patches ensemble method as proposed by:

> Heitor Murilo Gomes, Jesse Read, Albert Bifet.
    Streaming Random Patches for Evolving Data Stream Classification.
    IEEE International Conference on Data Mining (ICDM), 2019.

---

Checklist

- [x] Code complies with PEP-8 and is consistent with the framework.
- [x] Code is properly documented.
- [x] Tests are included for new functionality or updated accordingly.
- [x] Travis CI build passes with no errors.
- [x] Test Coverage is maintained (threshold is -0.2%).
- [x] Files changed (update, add, delete) are in the PR's scope (no extra files are included).

### CODE REVIEW

**Comment by hmgomes** [comment]
Hi Jacob, 

In my preliminary experiments I found that issue with WaveformGenerator. I could not get a broadcast error using a different learner (i.e. HoeffdingTreeClassifier), so I believe the problem is within the SRP implementation. To replicate the error:

from skmultiflow.data import WaveformGenerator
from skmultiflow.meta import StreamingRandomPatchesClassifier
from skmultiflow.evaluation import EvaluatePrequential
seed = 1
seed_stream = 0
evaluator = EvaluatePrequential(show_plot=False,
                                pretrain_size=200,
                                max_samples=20000)
srp = StreamingRandomPatchesClassifier(random_state=seed, n_estimators=2)
waveformStream = WaveformGenerator(random_state=seed_stream)
evaluator.evaluate(model=srp, stream=waveformStream)

It doesn't happen if we use a HoeffdingTreeClassifier or the AdaptiveRandomForest algorithm

from skmultiflow.data import WaveformGenerator
from skmultiflow.trees import HoeffdingTreeClassifier
from skmultiflow.evaluation import EvaluatePrequential
seed = 1
seed_stream = 0
evaluator = EvaluatePrequential(show_plot=False,
                                pretrain_size=200,
                                max_samples=20000)
ht = HoeffdingTreeClassifier()
waveformStream = WaveformGenerator(random_state=seed_stream)
evaluator.evaluate(model=ht, stream=waveformStream)

**Comment by jacobmontiel** [comment]
Nice catch! The error is triggered in a very specific scenario when members of the ensemble have observed different classes. The existing safe check only considered the case when a member predicts more classes than the ensemble. The reported bug corresponds to the case when a member predicts fewer classes than the ensemble. The fix is now in place.

File: src/skmultiflow/meta/streaming_random_patches.py [hunk 1]
```diff
@@ -0,0 +1,667 @@
...
+    def _reset_subset(self, n_features: int, random_state: np.random):
+        feature_indexes = None
+        if self.feature_indexes is not None:
+            k = len(self.feature_indexes)
+            feature_indexes = random_state.choice(range(n_features), k, replace=False)
+        return feature_indexes
+
+    def reset(self, n_features: int, n_samples_seen: int, random_state: np.random):
+        if not self.disable_background_learner and self._background_learner is not None:
+            self.base_estimator = deepcopy(self._background_learner.base_estimator)
```
Comment (smastelini) [inline src/skmultiflow/meta/streaming_random_patches.py:578]: I this case I believe that we could simply assign `self._background_learner.base_estimator` to `self.base_estimator`. What do you think? Is there a specific reason to copy the whole model?

File: tests/meta/test_streaming_random_patches.py [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+    y_pred = np.zeros(max_samples // n_wait, dtype=np.int)
+    y_true = np.zeros(max_samples // n_wait, dtype=np.int)
+    j = 0
+
+    for i in range(max_samples):
+        X, y = stream.next_sample()
+        # Test every n samples
+        if i % n_wait == 0:
+            y_pred[j] = int(learner.predict(X)[0])
+            y_true[j] = (y[0])
```
Comment (smastelini) [inline tests/meta/test_streaming_random_patches.py:114]: Is this statement missing a cast to `int`?

File: src/skmultiflow/meta/streaming_random_patches.py [hunk 3]
```diff
@@ -0,0 +1,667 @@
...
+from skmultiflow.utils import check_random_state, get_dimensions
+from skmultiflow.metrics import ClassificationPerformanceEvaluator
+
+
+class StreamingRandomPatchesClassifier(BaseSKMObject, ClassifierMixin, MetaEstimatorMixin):
+    """ Streaming Random Patches ensemble classifier.
+
+    Parameters
+    ----------
+    base_estimator: BaseSKMObject or sklearn.BaseObject, \
```
[the remark this answers is not in the document]
Comment (jacobmontiel) [reply src/skmultiflow/meta/streaming_random_patches.py:22]: The `\` is required so `sphinx` recognizes the next line as part of the definition of the parameter. Otherwise, it takes it as part of the description breaking the format of the documentation.

File: src/skmultiflow/meta/streaming_random_patches.py [hunk 4]
```diff
@@ -0,0 +1,667 @@
...
+                y_proba_temp = normalize(y_proba_temp, norm='l1')[0].copy()
+                acc = self.ensemble[i].performance_evaluator.accuracy_score()
+                if not self.disable_weighted_vote and acc > 0.0:
+                    y_proba_temp *= acc
+                # Check array length consistency
+                if len(y_proba_temp) != len(y_proba):
+                    if len(y_proba_temp) > len(y_proba):
+                        y_proba.resize((len(y_proba_temp), ), refcheck=False)
+                    else:
+                        y_proba_temp.resize((len(y_proba), ), refcheck=False)
```
[the remark this answers is not in the document]
Comment (jacobmontiel) [reply src/skmultiflow/meta/streaming_random_patches.py:318]: The base learner in the `AdaptiveRandomForestClassifier` is the `HoeffdingTreeClassifier` which can take `classes` to set the size of the probabilities array returned. In the `StreamingRandomPatchesClassifier` the base learner can be any type of estimator, so it can use an estimator that does not take into account `classes`.

File: src/skmultiflow/meta/streaming_random_patches.py [hunk 5]
```diff
@@ -0,0 +1,667 @@
...
+        self.performance_evaluator = performance_evaluator
+
+        # Store current model subspace representation of the original instances
+        self.feature_indexes = feature_indexes
+
+        # Drift detection
+        self.disable_background_learner = disable_background_learner
+        self.disable_drift_detector = disable_drift_detector
+        self.drift_detection_method = clone(drift_detection_method)   # type: BaseDriftDetector
+        self.warning_detection_method = clone(warning_detection_method)   # type: BaseDriftDetector
```
[the remark this answers is not in the document]
Comment (jacobmontiel) [reply src/skmultiflow/meta/streaming_random_patches.py:458]: I agree on the risk. However, `None` is not a valid option. `disable_drift_detection` is the correct way to achieve this. In this case, the user just needs to set the flag and leave the drift detector option to its default.

File: src/skmultiflow/meta/streaming_random_patches.py [hunk 6]
```diff
@@ -0,0 +1,667 @@
+from copy import deepcopy
+from typing import List, Optional
+from _collections import deque
```
[the remark this answers is not in the document]
Comment (jacobmontiel) [reply src/skmultiflow/meta/streaming_random_patches.py]: That is a typo. Fixed in 429b21a12408949d9bcb24ac8bb3e4e577f97224

File: src/skmultiflow/meta/streaming_random_patches.py [hunk 7]
```diff
@@ -0,0 +1,667 @@
...
+    """ Streaming Random Patches ensemble classifier.
+
+    Parameters
+    ----------
+    base_estimator: BaseSKMObject or sklearn.BaseObject, \
+        (default=HoeffdingTreeClassifier)
+        The base estimator.
+
+    n_estimators: int
+        Number of members in the ensemble.
```
[the remark this answers is not in the document]
Comment (jacobmontiel) [reply src/skmultiflow/meta/streaming_random_patches.py]: Sure. Added in 429b21a12408949d9bcb24ac8bb3e4e577f97224

File: src/skmultiflow/meta/streaming_random_patches.py [hunk 8]
```diff
@@ -0,0 +1,667 @@
...
+        Lambda value for bagging.
+
+    drift_detection_method: BaseDriftDetector, (default=ADWIN(delta=1e-5))
+        Drift detection method.
+
+    warning_detection_method: BaseDriftDetector, (default=ADWIN(delta=1e-4))
+        Warning detection method.
+
+    disable_weighted_vote: bool (default=False)
+        If True, uses weighted voting.
```
[the remark this answers is not in the document]
Comment (jacobmontiel) [reply src/skmultiflow/meta/streaming_random_patches.py]: There is actually a typo in the description. Changed in 429b21a12408949d9bcb24ac8bb3e4e577f97224 to `If True, disables weighted voting.`

File: src/skmultiflow/meta/streaming_random_patches.py [hunk 9]
```diff
@@ -0,0 +1,667 @@
...
+        ----------
+        X : numpy.ndarray of shape (n_samples, n_features)
+            The set of data samples to predict the class labels for.
+
+        Returns
+        -------
+        A numpy.ndarray with all the predictions for the samples in X.
+
+        """
+        n_samples, n_features = X.shape
```
[the remark this answers is not in the document]
Comment (jacobmontiel) [reply src/skmultiflow/meta/streaming_random_patches.py]: Changed in 429b21a12408949d9bcb24ac8bb3e4e577f97224

File: src/skmultiflow/meta/streaming_random_patches.py (hunk 1 above)
Comment (jacobmontiel) [reply src/skmultiflow/meta/streaming_random_patches.py]: You are right. Removed it in 429b21a12408949d9bcb24ac8bb3e4e577f97224

File: tests/meta/test_streaming_random_patches.py (hunk 2 above)
Comment (jacobmontiel) [reply tests/meta/test_streaming_random_patches.py]: Probably. I don't recall exactly. Just to be sure added in 429b21a12408949d9bcb24ac8bb3e4e577f97224

File: src/skmultiflow/meta/streaming_random_patches.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (smastelini) [reply src/skmultiflow/meta/streaming_random_patches.py:22]: Thank you for clarifying! Every new day I discover I know less than I though about programming XD

File: src/skmultiflow/meta/streaming_random_patches.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (smastelini) [reply src/skmultiflow/meta/streaming_random_patches.py]: Nice!

File: src/skmultiflow/meta/streaming_random_patches.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (smastelini) [reply src/skmultiflow/meta/streaming_random_patches.py:318]: Nice catch. Would in this case the probability vectors might become "unaligned"? (In case the base models observed different subsets of the classes)

File: src/skmultiflow/meta/streaming_random_patches.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (smastelini) [reply src/skmultiflow/meta/streaming_random_patches.py:458]: Agreed! It's a valid risk to take :)

File: src/skmultiflow/meta/streaming_random_patches.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (jacobmontiel) [reply src/skmultiflow/meta/streaming_random_patches.py:318]: The intention of the `resize` operation is to account for this case. For both cases, differences in other members of the ensemble, or in the current estimator.

**Comment by jacobmontiel** [comment]
## Correctness test

### Data
Data set: Electricity
`n_samples = 45312`
`n_features = 6`
`n_classes = 2`

### Estimator setup
`n_estimators = 10`
`random_stae = 1`
`training_method` = Random Patches (SRP), Resampling (BAG), Random Subspaces (SRS)

|                  | SRP   | BAG   | SRS   |
|------------------|-------|-------|-------|
| MOA              | 83.91 | 85.76 | 86.84 |
| scikit-multiflow | 84.90 | 86.21 | 84.66 |

Non-significant differences can be attributed to the different random number generators used.

File: src/skmultiflow/meta/streaming_random_patches.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (smastelini) [reply src/skmultiflow/meta/streaming_random_patches.py:318]: Thanks for the further clarification!
